sdk/metric/meter_test.go

```diff
@@ -474,3 +475,321 @@ func TestMetersProvideScope(t *testing.T) {
 	assert.NoError(t, err)
 	metricdatatest.AssertEqual(t, want, got, metricdatatest.IgnoreTimestamp())
 }
+
+func TestAttributeFilter(t *testing.T) {
+	one := 1.0
+	two := 2.0
+	testcases := []struct {
+		name       string
+		register   func(t *testing.T, mtr metric.Meter) error
+		wantMetric metricdata.Metrics
+	}{
+		{
+			name: "AsyncFloat64Counter",
+			register: func(t *testing.T, mtr metric.Meter) error {
+				ctr, err := mtr.AsyncFloat64().Counter("afcounter")
+				if err != nil {
+					return err
+				}
+				return mtr.RegisterCallback([]instrument.Asynchronous{ctr}, func(ctx context.Context) {
+					ctr.Observe(ctx, 1.0, attribute.Int("version", 1))
+					ctr.Observe(ctx, 2.0, attribute.Int("version", 2))
+				})
+			},
+			wantMetric: metricdata.Metrics{
+				Name: "afcounter",
+				Data: metricdata.Sum[float64]{
+					DataPoints: []metricdata.DataPoint[float64]{
+						{Value: 2.0}, // TODO: This should be 3.0.  It is a bug in the sum aggregator
```

Comment:
I think this bug is more subtle that just saying that it is in the sum aggregator (though, to be precise, it relates to the `precomputedSum`).

When a user passes a value to `Observe` it is interpreted as being the pre-compute cumulative sum value. If `Observe` is called twice in a callback and observed to be different values for the same attribute set, shouldn't the last value observed be used? It is the value the user is reporting the sum to be at. It seems to make sense in this situation that the last value is used.

However, the issue is the user didn't report the cumulative sum for the empty attribute set here, they report one for version `1` and one for `2`. The question becomes, how should precomputed sums be combined **when we know they are reported for different attribute sets**? This means that the solution to this issue is going to revolve around the [`filter`](https://github.com/open-telemetry/opentelemetry-go/blob/2fe8861a24e20088c065b116089862caf9e3cd8b/sdk/metric/internal/filter.go#L26) aggregation because that is the only place that know it is combining two distinct attribute sets (unless we redefine where attribute recording is done: https://github.com/open-telemetry/opentelemetry-go/issues/3030). The issue to resolve is it doesn't know it is combining them for the `precomputedSum` aggregator though.

Reply:
I took this approach because there was no special action required for multiple updates to aggregators.  It seems that we find ourselves with more complicated needs for this.

I don't think putting unique logic into the filter is the right idea.  Yes, this is the only place where a collision can be detected, but the filter doesn't understand the underlying aggregation.  So we would have to build special aggregators for each edge case.

This might mean we need to filter when the aggregator is read.  Most of the aggregators would use the same filter that is created here, but the Precomputed Sums would need something different.

Reply:
Can you open an issue in this project to track this bug and the conversation around it?